### DISCUSSION TITLE
Move network_time based on timer expiry

### DISCUSSION THREAD

**Discussion by awelzel** [zeek/zeek#4027] [Ideas]
Mostly making a note after discussing with @J-Gras about PCAP processing with long gaps.

As an example, the [testing/btest/Traces/ticks-dns-1hr.pcap](https://github.com/zeek/zeek/raw/refs/heads/master/testing/btest/Traces/ticks-dns-1hr.pcap) PCAP contains 10 hourly DNS packets. Reading this pcap and loading the following script results in 10 `tick()` executions. One before each packet.

```
# tick.zeek
event new_packet(c: connection, p: pkt_hdr)
        {
        print network_time(), "new_packet";
        }

event tick()
        {
        print network_time(), "tick()";
        schedule 10sec { tick() };
        }

event network_time_init()
        {
        print network_time(), "network_time_init";
        schedule 10sec { tick() };
        }
```

```
$ zeek -b -r testing/btest/Traces/ticks-dns-1hr.pcap testing/btest/tick.zeek
629503200.0, network_time_init
629503200.0, new_packet
629506800.0, tick()
629506800.0, new_packet
629510400.0, tick()
629510400.0, new_packet
629514000.0, tick()
629514000.0, new_packet
629517600.0, tick()
629517600.0, new_packet
629521200.0, tick()
629521200.0, new_packet
629524800.0, tick()
629524800.0, new_packet
629528400.0, tick()
629528400.0, new_packet
629532000.0, tick()
629532000.0, new_packet
629535600.0, tick()
629535600.0, new_packet
629535600.0, tick()
```

As can be seen, the first `tick()` event is running only when `network_time()` moved 1hr forward. Another expectation might be to have `tick()` execute many times (360) between two packets, and `network_time()` moving forward 10 seconds at a time.

The idea we had was to not set `network_time()` forward to packet time, but rather gradually to a timer's expiration time first, drain the queue, and repeat to allow further timers to be scheduled and expired. Eventually packet time would be reached.

This way, the 1 hour interval between two packets would be filled with some 360 tick() events that fire every 10 seconds and have  network_time() set appropriately, rather than just 1 tick() event for each packet.

Has this come up previously as a requirement / use-case? It feels "nicer" and makes some sense.

**Comment by vpax** [comment]
I don't know whether it's come up before. While I agree that the behavior can be unexpected, it also would appear to only matter in either offline PCAP situations, or very low traffic levels. I'm not aware of it messing up either of those enough to prioritize addressing this
